src/resizing.jl

```diff
@@ -23,111 +23,67 @@ function restrict{T,N}(A::AbstractArray{T,N}, dim::Integer)
-                    @inbounds @nloops $N i d->(d==1 ? (1:1) : (1:size(A,d))) d->(j_d = d==1 ? i_d+1 : i_d) begin
-                        nxt = convert(T, @nref $N A j)
-                        out[indx+=1] = half*(@nref $N A i) + quarter*nxt
-                        for k = 3:2:size(A,1)-2
-                            prv = nxt
-                            i_1 = k
-                            j_1 = k+1
-                            nxt = convert(T, @nref $N A j)
-                            out[indx+=1] = quarter*(prv+nxt) + half*(@nref $N A i)
-                        end
-                        i_1 = size(A,1)
-                        out[indx+=1] = quarter*nxt + half*(@nref $N A i)
-                    end
-                else
-                    strd = stride(out, dim)
-                    # Must initialize the i_dim==1 entries with zero
-                    @nexprs $N d->sz_d=d==dim?1:size(out,d)
-                    @nloops $N i d->(1:sz_d) begin
-                        (@nref $N out i) = zero(T)
-                    end
-                    stride_1 = 1
-                    @nexprs $N d->(stride_{d+1} = stride_d*size(out,d))
-                    @nexprs $N d->offset_d = 0
-                    ispeak = true
-                    @inbounds @nloops $N i d->(d==1?(1:1):(1:size(A,d))) d->(if d==dim; ispeak=isodd(i_d); offset_{d-1} = offset_d+(div(i_d+1,2)-1)*stride_d; else; offset_{d-1} = offset_d+(i_d-1)*stride_d; end) begin
-                        indx = offset_0
-                        if ispeak
-                            for k = 1:size(A,1)
-                                i_1 = k
-                                out[indx+=1] += half*(@nref $N A i)
-                            end
-                        else
-                            for k = 1:size(A,1)
-                                i_1 = k
-                                tmp = quarter*(@nref $N A i)
-                                out[indx+=1] += tmp
-                                out[indx+strd] = tmp
-                            end
-                        end
-                    end
-                end
+function restrict!{T,S,N}(out::AbstractArray{T,N}, A::AbstractArray{S}, dim)
+    # one_d is a tuple in which all elements are 0 expect for "dim"
+    #   e.g. for N=4 and dim=3: one_d=CartesianIndex((0,0,1,0))
+    one_d = CartesianIndex{N}(ntuple(d->d==dim?1:0, Val{N})::NTuple{N,Int})
+    # mlp is a tuple of multiplier in which all elements are 1 expect for "dim"
+    #   e.g. for N=4 and dim=3: mlp=(1,1,2,1)
+    mlp = ntuple(d->d==dim?2:1, Val{N})::NTuple{N,Int}
+    inds = indices(out)
+    min_d, max_d = extrema(inds[dim])
+    if isodd(length(indices(A,dim)))
+        half    = convert(eltype(T), 0.5)
+        quarter = convert(eltype(T), 0.25)
+        @inbounds for O in CartesianRange(inds)
+            # compute corresponding index in A
+            I = CartesianIndex{N}(map(*, O.I, mlp)) - one_d
+            if O[dim] == min_d
```

Comment:
Part of the performance of the previous version was avoiding branches in the inner loop. If you want to pursue this, one option would be to split the first dimension out from the rest.